### PR TITLE
Threading

### PR DESCRIPTION
**Pull request by ashamis**
- When using Raft we now assign a TLS connection to a thread and have that thread run the entire tx execution pipeline.
- Fix several other miscellaneous concurrency issues.

This is part of #774

### CODE REVIEW

File: src/crypto/symmkey.cpp [hunk 1]
```diff
@@ -14,49 +15,55 @@ namespace crypto
...
-    else if (n >= 192)
-    {
-      n_bits = 192;
-    }
-    else if (n >= 128)
-    {
-      n_bits = 128;
-    }
-    else
+    auto thread_count = enclave::ThreadMessaging::max_num_threads;
```
Comment (eddyashton) [inline src/crypto/symmkey.cpp]: `max_num_threads` should be constexpr, then `ctxs` can be `std::array<mbedtls_gcm_context, enclave::ThreadMessaging::max_num_threads>` (unless there's some reason I'm missing this actually needs to be `void*`?) and we don't need to resize.

File: src/ds/logger.h [hunk 2]
```diff
@@ -220,14 +220,14 @@ namespace logger
       return {};
     }
 
-    static inline std::vector<std::unique_ptr<AbstractLogger>>& loggers()
+    static inline std::vector<std::shared_ptr<AbstractLogger>>& loggers()
```
Comment (eddyashton) [inline src/ds/logger.h]: Why this change?

File: src/crypto/symmkey.cpp [hunk 3]
```diff
@@ -14,49 +15,55 @@ namespace crypto
...
+    ctxs.resize(thread_count);
+    for (uint16_t i = 0; i < thread_count; ++i)
     {
-      LOG_FATAL_FMT("Need at least {} bits, only have {}", 128, n);
-    }
+      void* ctx = new mbedtls_gcm_context;
```
Comment (achamayou) [inline src/crypto/symmkey.cpp]: Keep it typed?

**Comment by cimetrics** [comment]
threading_aaaa_5@[4585 aka 20200206.4](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=4585&view=results) vs master ewma over 30 builds from [4233](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=4233&view=results) to [4581](https://msrc-ccf.visualstudio.com/CCF/_build/results?buildId=4581&view=results)
images

**Comment by jumaffre** [comment]
AFAICS, there's still an issue in the encryptor where a key could be deleted (rollback or compact) while a transaction is being encrypted. As discussed, we should make sure we do not return a reference to a key in `get_encryption_key()` but a shared pointer? We can definitely sort this out in a separate PR though!
